assets/src/dashboard/app/views/storyAnimTool/index.js

```diff
@@ -0,0 +1,331 @@
+  const [activeStory, setActiveStory] = useState(null);
+  const [activeAnimation, setActiveAnimation] = useState({});
+  const [activePageIndex, setActivePageIndex] = useState(0);
+
+  const [selectedElementIds, setSelectedElementIds] = useState({});
+  const [isElementSelectable, setIsElementSelectable] = useState(false);
+
+  const {
+    actions: {
+      storyApi: { updateStory, fetchStories },
+    },
+    state: {
+      stories: { stories, storiesOrderById },
+    },
+  } = useContext(ApiContext);
+
+  useEffect(() => {
+    fetchStories({
+      searchTerm: '',
+      sortOption: STORY_SORT_OPTIONS.LAST_MODIFIED,
+      sortDirection: SORT_DIRECTION.DESC,
+      status: STORY_STATUS.ALL,
+      page: 1,
+      perPage: 25,
+    });
+  }, [fetchStories]);
+
+  const orderedStories = useMemo(() => {
+    return storiesOrderById.map((storyId) => {
+      return stories[storyId];
+    });
+  }, [stories, storiesOrderById]);
+
+  const saveActiveStoryUpdates = useCallback(() => {
+    const { content, story_data } = activeStory.originalStoryData;
+
+    updateStory({
+      ...activeStory,
+      content,
+      story_data: {
+        ...story_data,
+        pages: [...activeStory.pages],
+      },
+    });
+  }, [activeStory, updateStory]);
+
+  const handleSelectStory = useCallback(
+    (e) => {
+      if (activeStory) {
+        saveActiveStoryUpdates(activeStory);
+      }
+
+      const story = orderedStories.find(
+        (s) => s.id === parseInt(e.target.value)
+      );
+      setActiveStory(story);
+
+      //Deselect animation
+      setActiveAnimation({});
+    },
+    [activeStory, saveActiveStoryUpdates, orderedStories]
+  );
+
+  const handlePageNavClick = useCallback(
+    (offset) => {
+      const currentPageIndex = clamp(activePageIndex + offset, [
+        0,
+        activeStory.pages.length,
+      ]);
+
+      setActivePageIndex(currentPageIndex);
+    },
+    [activeStory, activePageIndex]
+  );
+
+  const handleSelectElement = useCallback(
+    (element) => {
+      if (selectedElementIds[element.id]) {
+        delete selectedElementIds[element.id];
+      } else {
+        selectedElementIds[element.id] = element.id;
+      }
+
+      setSelectedElementIds({ ...selectedElementIds });
+    },
+    [selectedElementIds]
+  );
+
+  const handleAddOrUpdateAnimation = useCallback(
+    (animation) => {
+      const animationWithTargets = {
+        ...animation,
+        targets: [...Object.values(selectedElementIds)],
+      };
+
+      const animations = activeStory.pages[activePageIndex].animations || [];
+      const index = animations.findIndex(
+        (a) => a.id === animationWithTargets.id
+      );
+
+      if (index < 0) {
+        // not found, push it to the end
+        animations.push(animationWithTargets);
+      } else {
+        // found, overwrite at index
+        animations[index] = animationWithTargets;
+      }
+
+      activeStory.pages[activePageIndex].animations = [...animations];
```

Comment:
Same here. I think this is mutating the source.

Reply:
I'm spreading the `activeStory` on `setActiveStory` so blowing away the original with a new, so I think this mutation is safe.  The only other thing I can think of is to copy `activeStory` at the top of the function, and then use that throughout 🤔, I'll go ahead and do that so it's more clear that things are being mutated.

Reply:
Okay, I pushed up an update 👍